pantheon-bundle/src/main/java/com/redhat/pantheon/servlet/ModuleRendererServlet.java

```diff
@@ -0,0 +1,105 @@
+
+import javax.servlet.Servlet;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.Writer;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.redhat.pantheon.conf.GlobalConfig.DEFAULT_MODULE_LOCALE;
+import static com.redhat.pantheon.model.module.ModuleVariant.DEFAULT_VARIANT_NAME;
+import static com.redhat.pantheon.servlet.ServletUtils.paramValue;
+import static com.redhat.pantheon.servlet.ServletUtils.paramValueAsBoolean;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Renders an HTML preview for a single module.
+ * To provide parameters to the asciidoc generation process, provide the parameters with their name prefixed
+ * with "ctx_".
+ *
+ * For example, if an asciidoc attribute of name 'product' needs to be passed, there will need to be a
+ * query parameter of name 'ctx_product' provided in the url.
+ */
+@Component(
+        service = Servlet.class,
+        property = {
+                Constants.SERVICE_DESCRIPTION + "=Servlet which transforms asciidoc content into html",
+                Constants.SERVICE_VENDOR + "=Red Hat Content Tooling team"
+        })
+@SlingServletResourceTypes(
+        resourceTypes = { "pantheon/module"},
+        methods = "GET",
+        extensions = "preview")
+public class ModuleRendererServlet extends SlingSafeMethodsServlet {
+
+    private final Logger log = LoggerFactory.getLogger(ModuleRendererServlet.class);
+
+    static final String PARAM_RERENDER = "rerender";
+    static final String PARAM_DRAFT = "draft";
+    static final String PARAM_LOCALE = "locale";
+    static final String PARAM_VARIANT = "variant";
+
+    private AsciidoctorService asciidoctorService;
+
+    @Activate
+    public ModuleRendererServlet(
+            @Reference AsciidoctorService asciidoctorService) {
+        this.asciidoctorService = asciidoctorService;
+    }
+
+    @Override
+    public void doGet(SlingHttpServletRequest request,
+            SlingHttpServletResponse response) throws IOException {
+        String locale = paramValue(request, PARAM_LOCALE, DEFAULT_MODULE_LOCALE.toString());
+        Boolean draft = paramValueAsBoolean(request, PARAM_DRAFT);
+        boolean reRender = paramValueAsBoolean(request, PARAM_RERENDER);
+        String variantName = paramValue(request, PARAM_VARIANT, DEFAULT_VARIANT_NAME);
+
+        Module module = request.getResource().adaptTo(Module.class);
+        Locale localeObj = LocaleUtils.toLocale(locale);
+
+        Optional<HashableFileResource> moduleVariantSource = null;
+
+        moduleVariantSource = module.moduleLocale(localeObj)
+                .traverse()
+                .toChild(ModuleLocale::source)
+                .toChild(draft ? SourceContent::draft : SourceContent::released)
+                .getAsOptional();
+
+
+        if(!moduleVariantSource.isPresent()) {
+            response.sendError(HttpServletResponse.SC_NOT_FOUND, (draft ? "Draft " : "Released ")
```

Comment:
String.format would be a better option than string concatenation.

Reply:
Better in what sense? String.format is my least favorite option.

If you are talking about "faster," that is not the case. String.format is the slowest option to use.
https://stackoverflow.com/a/2087703
https://redfin.engineering/java-string-concatenation-which-way-is-best-8f590a7d22a8

I would note that in older versions of java, string concatenation was handled poorly and was slow. However in newer versions of java, concatenation uses StringBuilder behind the scenes and is now fast and efficient.

If you have further thoughts then I would welcome the feedback, but otherwise I would prefer to leave this code block as-is.